### DISCUSSION TITLE
How to add a device and make it visible to fwupd to update firmware?

### DISCUSSION THREAD

**Discussion by niej** [fwupd/fwupd#5655] [Q&A] [answered]
Hi,

I have MTD device in system and need to update firmware on the MTD. Shall I add a configuration file or write a plugin to support this MTD device? I think a configuration file is reasonable, because it is a common requirement and interface is standard MTD device in Linux. But I cannot find an example.

Before I have my hardware in hand, I want to have some firmware update test on PC or raspberry pi device. Does USB disk or mmc is supported if just configuration file to add a firmware device in fwupd? If the answer is yes, is it possible to add offset parameter for MTD/MMC/USB-disk? Sometimes we just want to update a binary on a specific offset of storage.

Thanks!

**Comment by hughsie** [comment]
> or write a plugin to support this MTD device

It should work now with a new enough fwupd: https://github.com/fwupd/fwupd/tree/main/plugins/mtd -- what fwupd version are you testing with? And do you see the mtd device in `fwupdmgr get-devices`?

> is it possible to add offset parameter for MTD/MMC/USB-disk

We don't have that right now, but it should be easy enough to add. Do you need to erase, write and verify from a specific offset for a specific length? IIUC the kernel normally splits these up as different MTD nodes (e.g. on OpenBMC I see a MTD for kernel, rootfs etc) but we could certainly specify the offset+length in the quirk file too.

What hardware is this out of interest -- if you can share?

**Reply by niej** [reply]
I am using version: 1.7.5 for test on PC with adding below quirk file with referencing plugins/vli/vli-lenovo.quirk. fwupmgr cannot find my USB disk in this way. I can port latest fwupd to openbmc later. Testing with USB-disk on PC is more convenient for now. I assume discovering device is a mature feature on 1.7.5, is this right?

```$ cat /usr/share/fwupd/quirks.d/my-usb.quirk 
[USB\VID_058f&PID_6387]
Plugin = dfu
```

 I already read the README in plugins/mtd and section device-identification in tutorial, but still unclear how to add a new device for fwupd. I guess below quirk file is not enough, guid may be needed for . Meta.xml file only provides information for new firmware, not related to fwupd device discovering, right?


Sorry to confuse you on firmware on specific offset on storage. Only for MMC/USB-disk, I may need to write on specific offset.

**Reply by hughsie** [reply]
> Testing with USB-disk on PC is more convenient

Is `VID:058F PID:6387` a DFU device? You also need to have it in uppercase -- that's not going to work -- e.g. you want `USB\VID_058F&PID_6387` if the device actually supports the DFU protocol.

> still unclear how to add a new device for fwupd

Do you actually get /dev/mtdX devices on the device you're testing with?

> Meta.xml file only provides information for new firmware, not related to fwupd device discovering, right

Right.

> Only for MMC/USB-disk, I may need to write on specific offset.

Hang on a minute; do you mean MMC or MTD? MTD is exposed as a memory device and written to using mtd-specific tools. MMC is a mass storage controller and USB is a connection protocol. How do you update without using fwupd? Can you give an exact example please.

**Reply by niej** [reply]
My target is MTD, but I do not have hardware yet, so no /dev/mtdX yet. I want to test on PC with a USB disk for fwupd firmware update to get familiar with fwupd. USB disk is not a dfu device surely, I do not know which plugin is feasible for it, and thought DFU shall make it visible to fwupd at least. If USB disk is not possible, I plan to try mmc on raspberry pi. Do you have any suggestion on such a test?

**Reply by hughsie** [reply, marked as answer]
> I want to test on PC with a USB disk for fwupd firmware update

The mechanisms to update a USB device and an MTD device are completely different unfortunately. 

> DFU shall make it visible to fwupd at least

No, but you can use the "test" plugin in fwupd if you just want something to play with.

> I plan to try mmc on raspberry pi

If you mean MTD, I think this is a much more representative test. MMC and MTD are different plugins in fwupd.

**Comment by niej** [comment]
Thanks for the suggestion! What is missing with below setup? The test device cannot be found with ```fwupdmgr get-devices``` yet.

```
$ lsusb | grep 058
Bus 001 Device 008: ID 058f:6387 Alcor Micro Corp. Flash Drive

$ cat /usr/share/fwupd/quirks.d/my-test.quirk 
[USB\VID_058f&PID_6387]
Plugin = TEST
```

**Comment by niej** [comment]
Maybe we can add a segment in tutorial to explain how the devices are explored by fwupd. After reading related code and play on qemu quanta-gbs-bmc platform, I find that the mtd device is found by fwupd with parsing /dev/mtd* automatically. No extra steps are needed to make fwupd to be aware that mtd device is available to update firmware. Surely, this is true to other type of device, but it is not clear to new comer.

**Reply by hughsie** [reply]
Sure we could do, but please could you open a pull request with suggestions please.
